File: huffman-coding/huffman-coding.js
var input = '';

process.stdin.on('data', function(chunk) {
  input += chunk;
});

process.stdin.on('end', function() {
  main(input);
});

var EOF = String.fromCharCode(255);

function main(input) {
  input += EOF;

  var table = createTable(input);
  var encoded = encode(input, table);
  var decoded = decode(encoded, table);

  process.stdout.write(decoded);
}

function createTable(input) {
  var symbols = {}

  input.split('').forEach(function(c) {
    symbols[c] = symbols[c] || 0;
    symbols[c]++;
  });

  var chars = Object.keys(symbols);
  var nodes = chars.map(function(c) {
    return new Node(symbols[c], c);
  }).sort(function(a, b) {
    return a.weight < b.weight ? -1 : 1;
  });

  while (nodes.length > 1) {
    var a = nodes.shift();
    var b = nodes.shift();
    var node = new Node(a.weight + b.weight);
    node.left = a;
    node.right = b;

    var insertAt;
    nodes.forEach(function(n, i) {
      insertAt = i;
      return n.weight > node.weight;
    })
    nodes.splice(insertAt, 0, node);
  }

  var tree = nodes.shift();

  var table = {};
  chars.forEach(function(c) {
    table[c] = tree.getCode(c);
  });

  return table;
}

function encode(data, table) {
  var code = data.split('').map(function(c) {
    return table[c];
  }).join('');

  code = rpad(code, Math.ceil(code.length / 8) * 8);

  var arr = code.match(/.{1,8}/g).map(function(c) {
    return parseInt(c, 2);
  });

  return new Buffer(arr);
}

function decode(buf, table) {
  var invertedTable = invert(table);

  var code = '';
  for (var i = 0, len = buf.length; i < len; i++) {
    code += lpad(buf[i].toString(2), 8);
  }

  var decoded = '';
  var bits = '';
  code.split('').some(function(bit) {
    bits += bit;
    var c = invertedTable[bits]
    if (c === EOF) return true;
    if (c) {
      decoded += c;
      bits = '';
    }
  });

  return decoded;
}

function invert(obj) {
  var inverted = {};
  Object.keys(obj).forEach(function(k) {
    inverted[obj[k]] = k;
  });
  return inverted;
}

function lpad(str, width) {
  if (str.length >= width) return str;
  return new Array(width - str.length + 1).join('0') + str;
}

function rpad(str, width) {
  if (str.length >= width) return str;
  return str + new Array(width - str.length + 1).join('0');
}

function Node(weight, symbol) {
  this.weight = weight;
  this.symbol = symbol;
  this.left = null;
  this.right = null;
}

Node.prototype.getCode = function(symbol, prefix) {
  if (this.symbol === symbol) {
    return prefix;
  }

  prefix = prefix || '';

  if (this.left) {
    var code = this.left.getCode(symbol, prefix + '0');
    if (code) return code;
  }

  if (this.right) {
    var code = this.right.getCode(symbol, prefix + '1');
    if (code) return code;
  }
};

File: huffman-coding/huffman-coding.hs
import qualified Control.Monad as Monad
import qualified Data.ByteString as ByteString
import qualified Data.Char as Char
import qualified Data.List as List
import qualified Data.Map as Map
import qualified Data.Maybe as Maybe
import qualified Data.Tuple as Tuple
import qualified Data.Ord as Ord
import qualified Data.Word as Word
import qualified Numeric

data Node =
  LeafNode { weight :: Int, symbol :: Char } |
  InternalNode { weight :: Int, left :: Node, right :: Node }
  deriving (Show, Eq, Ord)

eof = Char.chr 255

createTable :: String -> [(Char, String)]
createTable = generateCodes
  . head
  . until (null . tail) insertInternalNode
  . List.sortBy compareWeight
  . map (\x -> LeafNode { weight = length x, symbol = head x })
  . List.group
  . List.sort

compareWeight :: Node -> Node -> Ordering
compareWeight = Ord.comparing weight

insertInternalNode :: [Node] -> [Node]
insertInternalNode (x1:x2:xs) =
  List.insertBy compareWeight InternalNode {
    weight = weight x1 + weight x2,
    left = x1,
    right = x2
  } xs

generateCodes :: Node -> [(Char, String)]
generateCodes LeafNode { symbol = s } = [(s, "")]
generateCodes InternalNode { left = l, right = r } =
  generateCodes' '0' l ++ generateCodes' '1' r

generateCodes' :: Char -> Node -> [(Char, String)]
generateCodes' p n = map (\(a, b) -> (a, p:b)) (generateCodes n)

encode :: [(Char, String)] -> String -> Maybe ByteString.ByteString
encode t = fmap ByteString.pack
  . Monad.join
  . fmap (mapM (readBin . rpad 8) . chunksOf 8 . concat)
  . mapM (`Map.lookup` m)
  where m = Map.fromList t

decode :: [(Char, String)] -> ByteString.ByteString -> Maybe String
decode t = parse m
  . concat
  . map (lpad 8 . showBin)
  . ByteString.unpack
  where m = Map.fromList . map Tuple.swap $ t

lpad :: Int -> String -> String
lpad n s
  | l < n = replicate (n - l) '0' ++ s
  | otherwise = s
  where l = length s

rpad :: Int -> String -> String
rpad n s
  | l < n = s ++ replicate (n - l) '0'
  | otherwise = s
  where l = length s

chunksOf :: Int -> [a] -> [[a]]
chunksOf _ [] = []
chunksOf n xs = a:(chunksOf n b)
  where (a, b) = splitAt n xs

readBin :: String -> Maybe Word.Word8
readBin = fmap fst
  . Maybe.listToMaybe
  . Numeric.readInt 2 (`elem` "01") Char.digitToInt

showBin :: Word.Word8 -> String
showBin n = Numeric.showIntAtBase 2 Char.intToDigit n ""

parse :: Map.Map String Char -> String -> Maybe String
parse m x = fmap init
  . fst
  . until' (fmap (elem eof) . fst)
    (\(a, b) -> let (c, d) = parse'' b in ((\a b -> a ++ [b]) <$> a <*> c, d))
  $ (Just "", x)
  where parse'' = parse' m

parse' :: Map.Map String Char -> String -> (Maybe Char, String)
parse' m x = (\(a, b) -> (Map.lookup a m, b))
  . until ((`Map.member` m) . fst) (\(a, b) -> (a ++ [head b], tail b))
  $ ("", x)

until' :: (a -> Maybe Bool) -> (a -> a) -> a -> a
until' p f = go
  where
    go x = case p x of
      Just True -> x
      _ -> go (f x)

main = do
  input <- getContents
  let input' = input ++ [eof]
  let table = createTable input'
  let encoded = encode table input'
  let decoded = decode table $ Maybe.fromJust encoded
  putStr $ Maybe.fromJust decoded
